sycl/source/detail/graph_impl.cpp

```
@@ -176,18 +219,26 @@ bool graph_impl::clear_queues() {
 
 // Check if nodes are empty and if so loop back through predecessors until we
 // find the real dependency.
-void find_real_deps(std::vector<pi_ext_sync_point> &Deps,
-                    std::shared_ptr<node_impl> CurrentNode) {
+void exec_graph_impl::find_real_deps(std::vector<pi_ext_sync_point> &Deps,
+                                     std::shared_ptr<node_impl> CurrentNode) {
   if (CurrentNode->is_empty()) {
     for (auto &N : CurrentNode->MPredecessors) {
       auto NodeImpl = N.lock();
       find_real_deps(Deps, NodeImpl);
     }
   } else {
-    // Check if the dependency has already been added.
-    if (std::find(Deps.begin(), Deps.end(), CurrentNode->MPiSyncPoint) ==
-        Deps.end())
-      Deps.push_back(CurrentNode->MPiSyncPoint);
+    // Verify that the sync point has actually been set for this node.
```

Is there a case where we need these verifications or is this purely for the development?

---

Hmm, I think it's a bit of both really. I'd say this is a fairly standard way of getting an element from a map, since `operator[]` often has unintended behaviour. The alternative is `unordered_map::at()` but that throws a C++ exception that may get swallowed or rethrown, and it's not really a recoverable error.

It is certainly useful for development as you say, and seeing such asserts is fairly common in the runtime from what I've seen. Plus the asserts can always be stripped out when building (I'm unsure if the official releases do this though).